src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs

```diff
@@ -337,6 +337,11 @@ private void ProcessExpressionParameter()
             if (ExcludeProperty != null)
             {
                 _exclusionFilter = new MshExpressionFilter(ExcludeProperty);
+                if ((Property == null) || (Property.Length == 0))
+                {
```

Comment:
I'd be in favour of non-breaking change, so I agree with pseudo-code @daxian-dbw proposed.
Although it's still far from the clean design, it's a) non-breaking for existing legit usages b) better than what we currently have. So I'm considering it a reasonable step in the right direction.